grails-app/services/org/pih/warehouse/core/LocationService.groovy

```diff
@@ -718,4 +724,83 @@ class LocationService {
+        } else if (params.locationNumber) {
+            location = Location.findByNameOrLocationNumber(params.name, params.locationNumber)
+        } else {
+            location = Location.findByName(params.name)
+        }
+        if (!location) {
+            location = new Location()
+        }
+
+        location.name = params.name
+        location.active = Boolean.valueOf(params.active)
+        location.locationNumber = params.locationNumber
+        location.locationGroup = params.locationGroup ? LocationGroup.findByName(params.locationGroup) : null
+        location.parentLocation = params.parentLocation ? Location.findByNameOrLocationNumber(params.parentLocation, params.parentLocation) : null
+        location.organization = params.organization ? Organization.findByCodeOrName(params.organization, params.organization) : null
+        location.address = createOrUpdateAddress(params, location?.address?.id)
+        if (!location.inventory && !location.parentLocation) {
+            location.inventory = new Inventory(['warehouse': location])
+        }
+        def locationType = null
+        if (params.locationType) {
+            String locationTypeName = LocalizationUtil.getDefaultString(params.locationType as String)
+            // TODO: Replace with a single GORM .find with Closure when in Grails 3 (available since Grails 2.0)
+            LocationType matchedLocationType = LocationType
+                    .findAllByNameLike(locationTypeName + "%")
+                    .find{ LocalizationUtil.getDefaultString(it.name) == locationTypeName}
+            locationType = matchedLocationType
+        }
+
+        def currentLocationType = location.locationType
+
+        //Do not allow to change internal type location to non-internal
+        if (((currentLocationType?.isInternalLocation() || currentLocationType?.isZone()) && (locationType?.isInternalLocation() || locationType?.isZone())) || !(currentLocationType?.isInternalLocation() || currentLocationType?.isZone())) {
+            location.locationType = locationType
+            location.supportedActivities = currentLocationType && locationType == currentLocationType ? location.supportedActivities : location.supportedActivities?.clear()
+        }
+
+        // Add required association to organization for depots and suppliers
+        if (!(location.locationType?.isInternalLocation() || location.locationType?.isZone()) && !location.organization) {
+            def locationCode = identifierService.generateOrganizationIdentifier(params.organization)
+            Organization organization = (location.locationType?.locationTypeCode == LocationTypeCode.SUPPLIER) ?
+                    organizationService.findOrCreateSupplierOrganization(params?.organization, locationCode) :
+                    organizationService.findOrCreateOrganization(params?.organization, locationCode)
+
+            location.organization = organization
+        }
+
+        return location
```

Comment:
I can see some inconsistency - as for address, we persist it in the `createOrUpdateAddress` method (which is fine in my opinion), but we don't persist the location here, but we return it unpersisted. Shouldn't we standardize it?